// countup.js-angular2/__ivy_ngcc__/fesm5/countup.js-angular2.js
import { __assign, __decorate, __metadata, __param } from 'tslib';
import { Input, Output, HostListener, Directive, Inject, PLATFORM_ID, ElementRef, NgZone, EventEmitter, NgModule } from '@angular/core';
import { CountUp } from 'countup.js';
import { isPlatformBrowser } from '@angular/common';

import * as ɵngcc0 from '@angular/core';
var CountUpDirective = /** @class */ (function () {
    function CountUpDirective(el, zone, platformId) {
        this.el = el;
        this.zone = zone;
        this.platformId = platformId;
        this.options = {};
        this.reanimateOnClick = true;
        this.complete = new EventEmitter();
    }
    // Re-animate if preference is set.
    CountUpDirective.prototype.onClick = function () {
        if (this.reanimateOnClick) {
            this.animate();
        }
    };
    CountUpDirective.prototype.ngOnChanges = function (changes) {
        // don't animate server-side (universal)
        if (!isPlatformBrowser(this.platformId)) {
            return;
        }
        if (changes.endVal && changes.endVal.currentValue !== undefined) {
            if (this.previousEndVal !== undefined) {
                this.options = __assign({}, this.options, { startVal: this.previousEndVal });
            }
            this.countUp = new CountUp(this.el.nativeElement, this.endVal, this.options);
            this.animate();
            this.previousEndVal = this.endVal;
        }
    };
    CountUpDirective.prototype.animate = function () {
        var _this = this;
        this.zone.runOutsideAngular(function () {
            _this.countUp.reset();
            _this.countUp.start(function () {
                _this.zone.run(function () {
                    _this.complete.emit();
                });
            });
        });
    };
    __decorate([
        Input('countUp'),
        __metadata("design:type", Number)
    ], CountUpDirective.prototype, "endVal", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], CountUpDirective.prototype, "options", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], CountUpDirective.prototype, "reanimateOnClick", void 0);
    __decorate([
        Output(),
        __metadata("design:type", Object)
    ], CountUpDirective.prototype, "complete", void 0);
    __decorate([
        HostListener('click'),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", void 0)
    ], CountUpDirective.prototype, "onClick", null);
    CountUpDirective = __decorate([ __param(2, Inject(PLATFORM_ID)),
        __metadata("design:paramtypes", [ElementRef,
            NgZone,
            Object])
    ], CountUpDirective);
CountUpDirective.ɵfac = function CountUpDirective_Factory(t) { return new (t || CountUpDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.NgZone), ɵngcc0.ɵɵdirectiveInject(PLATFORM_ID)); };
CountUpDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: CountUpDirective, selectors: [["", "countUp", ""]], hostBindings: function CountUpDirective_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("click", function CountUpDirective_click_HostBindingHandler() { return ctx.onClick(); });
    } }, inputs: { options: "options", reanimateOnClick: "reanimateOnClick", endVal: ["countUp", "endVal"] }, outputs: { complete: "complete" }, features: [ɵngcc0.ɵɵNgOnChangesFeature()] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(CountUpDirective, [{
        type: Directive,
        args: [{
                selector: '[countUp]'
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ɵngcc0.NgZone }, { type: Object, decorators: [{
                type: Inject,
                args: [PLATFORM_ID]
            }] }]; }, { options: [{
            type: Input
        }], reanimateOnClick: [{
            type: Input
        }], complete: [{
            type: Output
        }], onClick: [{
            type: HostListener,
            args: ['click']
        }], endVal: [{
            type: Input,
            args: ['countUp']
        }] }); })();
    return CountUpDirective;
}());

var CountUpModule = /** @class */ (function () {
    function CountUpModule() {
    }
CountUpModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: CountUpModule });
CountUpModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function CountUpModule_Factory(t) { return new (t || CountUpModule)(); }, imports: [[]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(CountUpModule, { declarations: [CountUpDirective], exports: [CountUpDirective] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(CountUpModule, [{
        type: NgModule,
        args: [{
                imports: [],
                declarations: [CountUpDirective],
                exports: [CountUpDirective]
            }]
    }], function () { return []; }, null); })();
    return CountUpModule;
}());

/*
 * Public API Surface of count-up
 */

/**
 * Generated bundle index. Do not edit.
 */

export { CountUpModule, CountUpDirective as ɵa };
